src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -37,82 +37,34 @@ public Checksum(HashData hash)
-                throw new ArgumentException($"checksum must be equal or bigger than the hash size: {HashSize}", nameof(checksum));
-            }
-
-            return FromWorker(checksum);
-        }
+            => From(checksum.AsSpan());
 
         /// <summary>
-        /// Create Checksum from given byte array. if byte array is bigger than
-        /// <see cref="HashSize"/>, it will be truncated to the size
+        /// Create Checksum from given byte array. if byte array is bigger than <see cref="HashSize"/>, it will be
+        /// truncated to the size.
         /// </summary>
         public static Checksum From(ImmutableArray<byte> checksum)
-        {
-            if (checksum.Length == 0)
-            {
-                return Null;
-            }
-
-            if (checksum.Length < HashSize)
-            {
-                throw new ArgumentException($"{nameof(checksum)} must be equal or bigger than the hash size: {HashSize}", nameof(checksum));
-            }
-
-            using var pooled = SharedPools.ByteArray.GetPooledObject();
-            var bytes = pooled.Object;
-            checksum.CopyTo(sourceIndex: 0, bytes, destinationIndex: 0, length: HashSize);
-
-            return FromWorker(bytes);
-        }
+            => From(checksum.AsSpan());
 
-        public static Checksum FromSerialized(byte[] checksum)
+        public static Checksum From(ReadOnlySpan<byte> checksum)
         {
             if (checksum.Length == 0)
-            {
                 return Null;
-            }
-
-            if (checksum.Length != HashSize)
-            {
-                throw new ArgumentException($"{nameof(checksum)} must be equal to the hash size: {HashSize}", nameof(checksum));
-            }
 
-            return FromWorker(checksum);
-        }
+            if (checksum.Length < HashSize)
+                throw new ArgumentException($"checksum must be equal or bigger than the hash size: {HashSize}", nameof(checksum));
 
-        private static unsafe Checksum FromWorker(byte[] checksum)
-        {
-            fixed (byte* data = checksum)
-            {
-                // Avoid a direct dereferencing assignment since sizeof(HashData) may be greater than HashSize.
-                //
-                // ex) "https://bugzilla.xamarin.com/show_bug.cgi?id=60298" - LayoutKind.Explicit, Size = 12 ignored with 64bit alignment
-                // or  "https://github.com/dotnet/roslyn/issues/23722" - Checksum throws on Mono 64-bit
-                return new Checksum(HashData.FromPointer((HashData*)data));
-            }
+            Contract.ThrowIfFalse(MemoryMarshal.TryRead(checksum, out HashData hash));
+            return new Checksum(hash);
         }
 
         public bool Equals(Checksum other)
         {
-            if (other == null)
-            {
-                return false;
-            }
-
-            return _checksum == other._checksum;
+            return other != null && _checksum == other._checksum;
```

Comment:
```suggestion
            return _checksum == other?._checksum;
```

Reply:
i thought about thsi, but i was under teh impression this was actuall sometimes slower due to the intro of nullable.

Reply:
You are correct. Seems we could do a much better job here.

```
using System.Runtime.CompilerServices;
using System.Threading;

public class C {
    public bool M1(CancellationToken token1, StrongBox<CancellationToken> token2) {
        return token2 != null && token1 == token2.Value;
    }

    public bool M2(CancellationToken token1, StrongBox<CancellationToken> token2) {
        return token1 == token2?.Value;
    }
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHABgAIcCA6AJQFcA7GNMOWgYQHswAHTIgDKiAG5oAxnCgBuUhWp0AKgAsEcCABM0zAObySCgMzU8lLpQDepSneqngPHhkoBZAgAouEZtIwYEGw8zMo8ANZwzJQwEVEESJTCMAghegBCPAAeADw+fnABQWghYZHMAHwxccx4AJTWtvbNOADs1eXmAIQAvJTMjAGUAGTDHfGUPX2xnbQAahAYjHCGzQC+pE12OI7Orm543r7+gcGhNePMCUkpaZm5+SfFpTVVM1H1jSTNLe3vV5NpjU8AB+eaLZarewbEhrIA

Reply:
@jaredpar is there a tracking item about potentially detecting and optimizing this idiom?  the latter code is arguably clearer (it is to me at least), but seems to generate a lot more heavyweight code for the task.  If not, i can open an issue.